# Pandas

# Introduction

 **Pandas** is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
 built on top of the Python programming language.

# Getting Started

To use pandas, you'll typically start with the following line of code.

In [1]:
import pandas as pd

# Creating data
There are two core objects in pandas: 
-  **DataFrame** 
-  **Series** 

### Dataframe

A DataFrame is a **table**. It contains an array of individual entries, each of which has a certain value. 
<br /> Each entry corresponds to a row (or record) and a column.

In [2]:
pd.DataFrame({'Roll No': [1, 3], 'Marks': [50, 87]})

Roll No  Marks
0        1     50
1        3     87

It can contain any types of data, in any tabular format as well.

In [3]:
pd.DataFrame({'Reviewer 1': ['I liked it.', 'It was awful.'], 
              'Reviewer 2': ['Pretty good.', 'Average.']},
             index=['Product A', 'Product B'])

Reviewer 1    Reviewer 2
Product A    I liked it.  Pretty good.
Product B  It was awful.      Average.


# Series
A Series, by contrast, is a sequence of data values. 
<br/> If a DataFrame is a table, a Series is a list. 
And in fact you can create one with nothing more than a list:



In [4]:
pd.Series([1, 2, 3, 4, 5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

A Series is, in essence, **a single column of a DataFrame**. So you can assign row labels to the Series the same way as before, using an index parameter.
> However, a Series does not have a column name, it only has **one overall name** .

In [5]:
pd.Series([30, 35, 40], index=['Physics', 'Chemistry', 'Maths'], name='Student A')

Physics      30
Chemistry    35
Maths        40
Name: Student A, dtype: int64

Hence dataframe is just a bunch of series glued together.

# Reading values in pandas
Pandas can be used to read data from files , especially CSV formatted files.
> A CSV file is a table of values separated by commas. Hence the name: "Comma-Separated Values", or CSV.

For reading CSV file, `pd.read_csv()` can be used.


In [6]:
spaceship_data=pd.read_csv("../input/spaceship-titanic/train.csv")

We can use the `shape` attribute to check how large the resulting DataFrame is:

In [7]:
spaceship_data.shape

(8693, 14)

This means it have  8693 records in 14 different columns! <br/>
We can check the first 5 rows of the data using the ` head()` command.

In [8]:
spaceship_data.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

If i want to check at each individual columns, i can do the following: 

In [9]:
spaceship_data.CryoSleep

0       False
1       False
2       False
3       False
4       False
        ...  
8688    False
8689     True
8690    False
8691    False
8692    False
Name: CryoSleep, Length: 8693, dtype: object

# Indexing in Pandas
Most of the indexing in Dataframe is similar to the common python indexing operations. However Pandas provide extra accessor operators like:
- Index- based selection
- Label- based selection

### Index based selection
Selecting data based on its numerical position in the data. `iloc` is used for this.
<br/>To select the first row of data in a DataFrame, we may use the following:

In [10]:
spaceship_data.iloc[0]

PassengerId             0001_01
HomePlanet               Europa
CryoSleep                 False
Cabin                     B/0/P
Destination         TRAPPIST-1e
Age                        39.0
VIP                       False
RoomService                 0.0
FoodCourt                   0.0
ShoppingMall                0.0
Spa                         0.0
VRDeck                      0.0
Name            Maham Ofracculy
Transported               False
Name: 0, dtype: object

Or we can use more advanced indexing like selecting specific columns and rows using `iloc`.

In [11]:
# Shows only second and third entries of the data in the first column
spaceship_data.iloc[1:3, 0]

1    0002_01
2    0003_01
Name: PassengerId, dtype: object

### Label based selection
In this type, data index value is used, not the position of the data.
<br/> `loc` operator is commonly used for this.

In [12]:
# Listing down the first row valus of the specific column.
spaceship_data.loc[1, 'HomePlanet']

'Earth'

# Assigning values
Values can also be assigned to dataFrame much like other python operations.


In [13]:
spaceship_data['FoodCourt']=range(len(spaceship_data), 0, -1)
spaceship_data

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
0        0001_01     Europa     False     B/0/P    TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False     F/0/S    TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False     A/0/S    TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False     A/0/S    TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False     F/1/S    TRAPPIST-1e  16.0  False   
...          ...        ...       ...       ...            ...   ...    ...   
8688     9276_01     Europa     False    A/98/P    55 Cancri e  41.0   True   
8689     9278_01      Earth      True  G/1499/S  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False  G/1500/S    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False   E/608/S    55 Cancri e  32.0  False   
8692     9280_02     Europa     False   E/608/S    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0             0.0       8693           0.0     0.0     0.0    Maham Ofracculy   
1           109.0       8692          25.0   549.0    44.0       Juanna Vines   
2            43.0       8691           0.0  6715.0    49.0      Altark Susent   
3             0.0       8690         371.0  3329.0   193.0       Solam Susent   
4           303.0       8689         151.0   565.0     2.0  Willy Santantines   
...           ...        ...           ...     ...     ...                ...   
8688          0.0          5           0.0  1643.0    74.0  Gravior Noxnuther   
8689          0.0          4           0.0     0.0     0.0    Kurta Mondalley   
8690          0.0          3        1872.0     1.0     0.0       Fayey Connon   
8691          0.0          2           0.0   353.0  3235.0   Celeon Hontichre   
8692        126.0          1           0.0     0.0    12.0   Propsh Hontichre   

      Transported  
0           False  
1            True  
2           False  
3           False  
4            True  
...           ...  
8688        False  
8689        False  
8690         True  
8691        False  
8692         True  

[8693 rows x 14 columns]

# Summary Functions
Some of the summary  functions used in pandas are:
- **describe()**
- **mean()**
- **unique()**
- **value_counts()**

let's look at each  one of them briefly.

### Describe()
- it will give a  summary the content of the selected column of the DataFrame
- it is a type-aware property, it means its output changes with the type of data in the dataframe


In [14]:
spaceship_data.HomePlanet.describe()

count      8492
unique        3
top       Earth
freq       4602
Name: HomePlanet, dtype: object

In [15]:
spaceship_data.FoodCourt.describe()

count    8693.000000
mean     4347.000000
std      2509.597278
min         1.000000
25%      2174.000000
50%      4347.000000
75%      6520.000000
max      8693.000000
Name: FoodCourt, dtype: float64

### mean()
- produce mean of a particular column

In [16]:
spaceship_data.FoodCourt.mean()

4347.0

### median()
- produce median of a particular column

In [17]:
spaceship_data.FoodCourt.median()

4347.0

### unique()
- List the unique elements of a column

In [18]:
spaceship_data.HomePlanet.unique()

array(['Europa', 'Earth', 'Mars', nan], dtype=object)

### value_counts()
- To see a list of unique values and how often they occur in the dataset, we can use the `value_counts()` method

In [19]:
spaceship_data.HomePlanet.value_counts()

Earth     4602
Europa    2131
Mars      1759
Name: HomePlanet, dtype: int64

# Maps
In data science we often have a need for **creating new representations from existing data, or for transforming data** from the format it is in now to the format that we want it to be in later. Maps are what handle this work, making them extremely important for getting your work done!
<br /> The main functions used are :
- **map()**
- **apply()**


### map()
- The function you pass to **map()** should expect a single value from the Series (a point value, in the above example), and return a transformed version of that value.
- map() returns a new Series where all the values have been transformed by your function.

In [20]:
name_list=spaceship_data.HomePlanet.map({'Europa': 'Titan', 'Mars': 'Red Giant'})
print(name_list)

0       Titan
1         NaN
2       Titan
3       Titan
4         NaN
        ...  
8688    Titan
8689      NaN
8690      NaN
8691    Titan
8692    Titan
Name: HomePlanet, Length: 8693, dtype: object


### apply()
- apply() is the equivalent method if we want to transform a **whole DataFrame** *by calling a custom method on each row*.

In [21]:
def stars(row):
    if row.VIP == True:
        return "survivor"
    elif row.CryoSleep == True:
        return "survivor"
    else :
        return "Not Survived"
    
planet_ratings = spaceship_data.apply(stars, axis='columns')
print(planet_ratings)

0       Not Survived
1       Not Survived
2           survivor
3       Not Survived
4       Not Survived
            ...     
8688        survivor
8689        survivor
8690    Not Survived
8691    Not Survived
8692    Not Survived
Length: 8693, dtype: object


## Groupwise analysis
In terms of large amount of data, proper grouping of data is required for in depth analysis of the data.
<br/>The following functions are majorly used for the group wise analysis.
- **groupby()**
- **agg()**
- **idxmax()**

### groupby()
- grouping columns based on a metric

In [22]:
Martians = spaceship_data.groupby('VIP').size()
Martians

VIP
False    8291
True      199
dtype: int64

### agg()
- which lets you run a bunch of different functions on your DataFrame simultaneously

In [23]:
room_data = spaceship_data.groupby('RoomService').agg([min,max])
room_data

PassengerId          CryoSleep          Age          VIP         \
                    min      max       min    max   min   max    min    max   
RoomService                                                                   
0.0             0001_01  9280_01     False   True   0.0  79.0  False   True   
1.0             0066_01  9275_02     False  False  13.0  67.0  False   True   
2.0             0206_01  9239_01     False  False  13.0  61.0  False   True   
3.0             0286_01  9233_01     False  False  14.0  63.0  False   True   
4.0             0052_01  8834_01     False  False  13.0  50.0  False  False   
...                 ...      ...       ...    ...   ...   ...    ...    ...   
8209.0          7585_02  7585_02     False  False  39.0  39.0    NaN    NaN   
8243.0          8470_01  8470_01       NaN    NaN  31.0  31.0   True   True   
8586.0          9202_01  9202_01     False  False  26.0  26.0  False  False   
9920.0          5455_01  5455_01     False  False  13.0  13.0  False  False   
14327.0         4690_02  4690_02     False  False  27.0  27.0  False  False   

            FoodCourt       ShoppingMall              Spa           VRDeck  \
                  min   max          min      max     min      max     min   
RoomService                                                                  
0.0                 2  8693          0.0  23492.0     0.0  18572.0     0.0   
1.0                 7  8633          0.0   4790.0     0.0  13104.0     0.0   
2.0                33  8509          0.0   4509.0     0.0   5577.0     0.0   
3.0                37  8434          0.0   2908.0     0.0  10860.0     0.0   
4.0               420  8643          0.0   2574.0     0.0   4683.0     0.0   
...               ...   ...          ...      ...     ...      ...     ...   
8209.0           1575  1575        432.0    432.0  3560.0   3560.0    54.0   
8243.0            760   760          0.0      0.0  3220.0   3220.0  2285.0   
8586.0             67    67        114.0    114.0    29.0     29.0     4.0   
9920.0           3588  3588        180.0    180.0   130.0    130.0   571.0   
14327.0          4277  4277          0.0      0.0     1.0      1.0     0.0   

                     Transported         
                 max         min    max  
RoomService                              
0.0          20336.0       False   True  
1.0          12682.0       False   True  
2.0          14485.0       False   True  
3.0          12392.0       False   True  
4.0           3656.0       False   True  
...              ...         ...    ...  
8209.0          54.0       False  False  
8243.0        2285.0       False  False  
8586.0           4.0       False  False  
9920.0         571.0       False  False  
14327.0          0.0       False  False  

[1273 rows x 18 columns]

### idxmax()
- The idxmax() method returns a Series with the index of the maximum value for each column. By specifying the column axis ( axis='columns' ).
- the idxmax() method returns a Series with the index of the maximum value for each row.

In [24]:
data = {
  "sales": [23, 34, 56],
  "age": [50, 40, 30]
}

df = pd.DataFrame(data)

print(df.idxmax())

sales    2
age      0
dtype: int64


# Data Types 
The data type for a column in a DataFrame or a Series is known as the **dtype.**
<br/> The funcitons related to Data types are:
- dtype
- dtypes
- astype()
- fillna()
- replace()
- sort_values()
- rename()
- rename_axis()


### dtype
- shows the data type of the column

In [25]:
spaceship_data.RoomService.dtype

dtype('float64')

### dtypes
- shows data type pf every column of the dataframe

In [26]:
spaceship_data.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt         int64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

### astype()
- It's possible to convert a column of one type into another wherever such a conversion makes sense by using the astype()

In [27]:
spaceship_data.HomePlanet.astype(str)

0       Europa
1        Earth
2       Europa
3       Europa
4        Earth
         ...  
8688    Europa
8689     Earth
8690     Earth
8691    Europa
8692    Europa
Name: HomePlanet, Length: 8693, dtype: object

### fillna()
- Filling the NaN value with the value we required
- we could fill each missing value with the first non-null value that appears sometime after the given record in the database. This is known as the backfill strategy.


In [28]:
spaceship_data.HomePlanet.fillna("NA")

0       Europa
1        Earth
2       Europa
3       Europa
4        Earth
         ...  
8688    Europa
8689     Earth
8690     Earth
8691    Europa
8692    Europa
Name: HomePlanet, Length: 8693, dtype: object

### replace()
- to replace a particular value with another in dataframe

In [29]:
spaceship_data.HomePlanet.replace("NA","Not Available")

0       Europa
1        Earth
2       Europa
3       Europa
4        Earth
         ...  
8688    Europa
8689     Earth
8690     Earth
8691    Europa
8692    Europa
Name: HomePlanet, Length: 8693, dtype: object

### sort_values()
- sort the data on a specific value

In [30]:
spaceship_data.HomePlanet.sort_values()

2765    Earth
3546    Earth
3545    Earth
3544    Earth
3543    Earth
        ...  
8515      NaN
8613      NaN
8666      NaN
8674      NaN
8684      NaN
Name: HomePlanet, Length: 8693, dtype: object

### rename()
- rename columns / index names
- rename() lets you rename index or column values by specifying a index or column keyword parameter, respectively. It supports a variety of input formats, but usually a Python dictionary is the most convenient. 

In [31]:
mall_value=spaceship_data.rename(columns={'ShoppingMall': 'Mall'})
mall_value.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt   Mall     Spa  VRDeck               Name  \
0          0.0       8693    0.0     0.0     0.0    Maham Ofracculy   
1        109.0       8692   25.0   549.0    44.0       Juanna Vines   
2         43.0       8691    0.0  6715.0    49.0      Altark Susent   
3          0.0       8690  371.0  3329.0   193.0       Solam Susent   
4        303.0       8689  151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

### rename_axis()
- can be used to index of a row or column quickly.

In [32]:
renameaxis=spaceship_data.rename_axis("data", axis='rows')
renameaxis.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
data                                                                     
0        0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
data                                                                            
0             0.0       8693           0.0     0.0     0.0    Maham Ofracculy   
1           109.0       8692          25.0   549.0    44.0       Juanna Vines   
2            43.0       8691           0.0  6715.0    49.0      Altark Susent   
3             0.0       8690         371.0  3329.0   193.0       Solam Susent   
4           303.0       8689         151.0   565.0     2.0  Willy Santantines   

      Transported  
data               
0           False  
1            True  
2           False  
3           False  
4            True

# Combining Operators
- concat()
- join()
- merge()

### concat()
- concating one dataframe on another dataframe

In [33]:
data1=spaceship_data.HomePlanet
data2=spaceship_data.Transported
pd.concat([data1,data2])

0       Europa
1        Earth
2       Europa
3       Europa
4        Earth
         ...  
8688     False
8689     False
8690      True
8691     False
8692      True
Length: 17386, dtype: object

### join()
- lsuffix and rsuffix parameters can be used as attributes with the join() function

In [34]:
data3=spaceship_data
data4=spaceship_data
data3.join(data4,lsuffix='_CAN', rsuffix='_UK')

PassengerId_CAN HomePlanet_CAN CryoSleep_CAN Cabin_CAN Destination_CAN  \
0            0001_01         Europa         False     B/0/P     TRAPPIST-1e   
1            0002_01          Earth         False     F/0/S     TRAPPIST-1e   
2            0003_01         Europa         False     A/0/S     TRAPPIST-1e   
3            0003_02         Europa         False     A/0/S     TRAPPIST-1e   
4            0004_01          Earth         False     F/1/S     TRAPPIST-1e   
...              ...            ...           ...       ...             ...   
8688         9276_01         Europa         False    A/98/P     55 Cancri e   
8689         9278_01          Earth          True  G/1499/S   PSO J318.5-22   
8690         9279_01          Earth         False  G/1500/S     TRAPPIST-1e   
8691         9280_01         Europa         False   E/608/S     55 Cancri e   
8692         9280_02         Europa         False   E/608/S     TRAPPIST-1e   

      Age_CAN VIP_CAN  RoomService_CAN  FoodCourt_CAN  ShoppingMall_CAN  ...  \
0        39.0   False              0.0           8693               0.0  ...   
1        24.0   False            109.0           8692              25.0  ...   
2        58.0    True             43.0           8691               0.0  ...   
3        33.0   False              0.0           8690             371.0  ...   
4        16.0   False            303.0           8689             151.0  ...   
...       ...     ...              ...            ...               ...  ...   
8688     41.0    True              0.0              5               0.0  ...   
8689     18.0   False              0.0              4               0.0  ...   
8690     26.0   False              0.0              3            1872.0  ...   
8691     32.0   False              0.0              2               0.0  ...   
8692     44.0   False            126.0              1               0.0  ...   

      Destination_UK  Age_UK VIP_UK  RoomService_UK FoodCourt_UK  \
0        TRAPPIST-1e    39.0  False             0.0         8693   
1        TRAPPIST-1e    24.0  False           109.0         8692   
2        TRAPPIST-1e    58.0   True            43.0         8691   
3        TRAPPIST-1e    33.0  False             0.0         8690   
4        TRAPPIST-1e    16.0  False           303.0         8689   
...              ...     ...    ...             ...          ...   
8688     55 Cancri e    41.0   True             0.0            5   
8689   PSO J318.5-22    18.0  False             0.0            4   
8690     TRAPPIST-1e    26.0  False             0.0            3   
8691     55 Cancri e    32.0  False             0.0            2   
8692     TRAPPIST-1e    44.0  False           126.0            1   

     ShoppingMall_UK  Spa_UK VRDeck_UK            Name_UK  Transported_UK  
0                0.0     0.0       0.0    Maham Ofracculy           False  
1               25.0   549.0      44.0       Juanna Vines            True  
2                0.0  6715.0      49.0      Altark Susent           False  
3              371.0  3329.0     193.0       Solam Susent           False  
4              151.0   565.0       2.0  Willy Santantines            True  
...              ...     ...       ...                ...             ...  
8688             0.0  1643.0      74.0  Gravior Noxnuther           False  
8689             0.0     0.0       0.0    Kurta Mondalley           False  
8690          1872.0     1.0       0.0       Fayey Connon            True  
8691             0.0   353.0    3235.0   Celeon Hontichre           False  
8692             0.0     0.0      12.0   Propsh Hontichre            True  

[8693 rows x 28 columns]

### merge()
- merging two dataframes

In [35]:
data3=spaceship_data
data4=spaceship_data
data3.merge(data4)

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
0        0001_01     Europa     False     B/0/P    TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False     F/0/S    TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False     A/0/S    TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False     A/0/S    TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False     F/1/S    TRAPPIST-1e  16.0  False   
...          ...        ...       ...       ...            ...   ...    ...   
8688     9276_01     Europa     False    A/98/P    55 Cancri e  41.0   True   
8689     9278_01      Earth      True  G/1499/S  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False  G/1500/S    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False   E/608/S    55 Cancri e  32.0  False   
8692     9280_02     Europa     False   E/608/S    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0             0.0       8693           0.0     0.0     0.0    Maham Ofracculy   
1           109.0       8692          25.0   549.0    44.0       Juanna Vines   
2            43.0       8691           0.0  6715.0    49.0      Altark Susent   
3             0.0       8690         371.0  3329.0   193.0       Solam Susent   
4           303.0       8689         151.0   565.0     2.0  Willy Santantines   
...           ...        ...           ...     ...     ...                ...   
8688          0.0          5           0.0  1643.0    74.0  Gravior Noxnuther   
8689          0.0          4           0.0     0.0     0.0    Kurta Mondalley   
8690          0.0          3        1872.0     1.0     0.0       Fayey Connon   
8691          0.0          2           0.0   353.0  3235.0   Celeon Hontichre   
8692        126.0          1           0.0     0.0    12.0   Propsh Hontichre   

      Transported  
0           False  
1            True  
2           False  
3           False  
4            True  
...           ...  
8688        False  
8689        False  
8690         True  
8691        False  
8692         True  

[8693 rows x 14 columns]